substrate/frame/staking/src/slashing.rs

```diff
@@ -609,26 +609,28 @@ pub fn do_slash<T: Config>(
 		};
 
 	let value = ledger.slash(value, T::Currency::minimum_balance(), slash_era);
+	if value.is_zero() {
+		// nothing to do
+		return
+	}
 
-	if !value.is_zero() {
+	// Skip slashing for virtual stakers. The pallets managing them should handle the slashing.
+	if !Pallet::<T>::is_virtual_staker(stash) {
 		let (imbalance, missing) = T::Currency::slash(stash, value);
 		slashed_imbalance.subsume(imbalance);
 
 		if !missing.is_zero() {
 			// deduct overslash from the reward payout
 			*reward_payout = reward_payout.saturating_sub(missing);
 		}
+	}
 
-		let _ = ledger
-			.update()
-			.defensive_proof("ledger fetched from storage so it exists in storage; qed.");
+	let _ = ledger
+		.update()
+		.defensive_proof("ledger fetched from storage so it exists in storage; qed.");
 
-		// trigger the event
-		<Pallet<T>>::deposit_event(super::Event::<T>::Slashed {
-			staker: stash.clone(),
-			amount: value,
-		});
-	}
+	// trigger the event
+	<Pallet<T>>::deposit_event(super::Event::<T>::Slashed { staker: stash.clone(), amount: value });
```

Comment:
This event will be deposited if it is a virtual staker with an amount that actually was not slashed here. Maybe we could return earlier if the ledger is virtual instead? Or throw a different event here than `Slashed`.

Reply:
From this pallet's point of view it is still slashed right? The staking ledger is slashed.

For virtual stakers, this staking pallet does not manage locks so it cannot slash but it sends a slashing event for lock managing pallet to apply slashes.

If I am too pedantic, I would say we can have different events for each type of staker, but I think it is an overkill. Also, slash of a virtual staker will emit two events, this and another by delegated staking pallet which is enough to know for a dapp that this is a virtual staker. Wdyt?